codec/src/main/java/io/netty/handler/codec/string/LineEncoder.java

```diff
@@ -0,0 +1,96 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.string;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.LineBasedFrameDecoder;
+import io.netty.handler.codec.MessageToMessageEncoder;
+import io.netty.util.internal.ObjectUtil;
+
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.util.List;
+
+/**
+ * Apply a line separator to the requested {@link String} and encode it into a {@link ByteBuf}.
+ * A typical setup for a text-based line protocol in a TCP/IP socket would be:
+ * <pre>
+ * {@link ChannelPipeline} pipeline = ...;
+ *
+ * // Decoders
+ * pipeline.addLast("frameDecoder", new {@link LineBasedFrameDecoder}(80));
+ * pipeline.addLast("stringDecoder", new {@link StringDecoder}(CharsetUtil.UTF_8));
+ *
+ * // Encoder
+ * pipeline.addLast("lineEncoder", new {@link LineEncoder}(LineSeparator.UNIX, CharsetUtil.UTF_8));
+ * </pre>
+ * and then you can use a {@link String} instead of a {@link ByteBuf}
+ * as a message:
+ * <pre>
+ * void channelRead({@link ChannelHandlerContext} ctx, {@link String} msg) {
+ *     ch.write("Did you say '" + msg + "'?");
+ * }
+ * </pre>
+ */
+@Sharable
+public class LineEncoder extends MessageToMessageEncoder<CharSequence> {
+
+    private final Charset charset;
+
+    private final byte[] lineSeparator;
+
+    /**
+     * Creates a new instance with the current system line separator and character set.
+     */
+    public LineEncoder() {
+        this(LineSeparator.DEFAULT, Charset.defaultCharset());
+    }
+
+    /**
+     * Creates a new instance with the specified line separator.
+     */
+    public LineEncoder(LineSeparator lineSeparator) {
+        this(lineSeparator, Charset.defaultCharset());
+    }
+
+    /**
+     * Creates a new instance with the specified character set.
+     */
+    public LineEncoder(Charset charset) {
+        this(LineSeparator.DEFAULT, charset);
+    }
+
+    /**
+     * Creates a new instance with the specified line separator and character set.
+     */
+    public LineEncoder(LineSeparator lineSeparator, Charset charset) {
+        ObjectUtil.checkNotNull(charset, "charset");
+        ObjectUtil.checkNotNull(lineSeparator, "lineSeparator");
+        this.charset = charset;
+        this.lineSeparator = lineSeparator.value().getBytes(charset);
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, CharSequence msg, List<Object> out) throws Exception {
+        ByteBuf buf = ByteBufUtil.encodeString(ctx.alloc(), CharBuffer.wrap(msg), charset);
```

Comment:
I think better would be to calculate the assumed length of the buffer before while taking the `lineSeperator` into account. Otherwise you may need to resize the buffer when writing it.